src/process_utility.c

```diff
@@ -970,6 +974,159 @@ process_index_end(Node *parsetree, CollectedCommand *cmd)
+ * involves taking exclusive locks on all tables for extensive periods of time,
+ * each subtable is clustered in its own transaction. This will release all
+ * locks on subtables once they are done.
+ */
+static bool
+process_cluster_start(Node *parsetree, ProcessUtilityContext context)
+{
+	ClusterStmt *stmt = (ClusterStmt *) parsetree;
+	Cache	   *hcache;
+	Hypertable *ht;
+
+	Assert(IsA(stmt, ClusterStmt));
+
+	/* If this is a re-cluster on all tables, there is nothing we need to do */
+	if (NULL == stmt->relation)
+		return false;
+
+	hcache = hypertable_cache_pin();
+	ht = hypertable_cache_get_entry_rv(hcache, stmt->relation);
+
+	if (NULL != ht)
+	{
+		bool		is_top_level = (context == PROCESS_UTILITY_TOPLEVEL);
+		Oid			index_relid;
+		List	   *chunk_indexes;
+		ListCell   *lc;
+		MemoryContext old,
+					mcxt;
+
+		if (!pg_class_ownercheck(ht->main_table_relid, GetUserId()))
+			aclcheck_error(ACLCHECK_NOT_OWNER, ACL_KIND_CLASS,
+						   get_rel_name(ht->main_table_relid));
+
+		/*
+		 * If CLUSTER is run inside a user transaction block; we bail out or
+		 * otherwise we'd be holding locks way too long.
+		 */
+		PreventTransactionChain(is_top_level, "CLUSTER");
+
+		if (NULL == stmt->indexname)
+			index_relid = find_clustered_index(ht->main_table_relid);
+		else
+			index_relid = get_relname_relid(stmt->indexname,
+									get_rel_namespace(ht->main_table_relid));
+
+		if (!OidIsValid(index_relid))
+		{
+			/* Let regular process utility handle */
+			cache_release(hcache);
+			return false;
+		}
+
+		/*
+		 * The list of chunks and their indexes need to be on a memory context
+		 * that will survive moving to a new transaction for each chunk
+		 */
+		mcxt = AllocSetContextCreate(PortalContext,
+									 "Hypertable cluster",
+									 ALLOCSET_DEFAULT_SIZES);
+
+		/*
+		 * Get a list of chunks and indexes that correspond to the
+		 * hypertable's index
+		 */
+		old = MemoryContextSwitchTo(mcxt);
+		chunk_indexes = chunk_index_get_mappings(ht, index_relid);
+		MemoryContextSwitchTo(old);
+
+		/* Commit to get out of starting transaction */
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		foreach(lc, chunk_indexes)
+		{
+			ChunkIndexMapping *cim = lfirst(lc);
+
+			/* Start a new transaction for each relation. */
+			StartTransactionCommand();
+			/* functions in indexes may want a snapshot set */
+			PushActiveSnapshot(GetTransactionSnapshot());
+
+			/*
+			 * We must mark each chunk index as clustered before calling
+			 * cluster_rel() because it expects indexes that need to be
+			 * rechecked (due to new transaction) to already have that mark
+			 * set
+			 */
+			chunk_index_mark_clustered(cim->chunkoid, cim->indexoid);
+
+			/* Do the job. */
+			cluster_rel(cim->chunkoid, cim->indexoid, true, stmt->verbose);
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+		}
+		/* Start a new transaction for the cleanup work. */
+		StartTransactionCommand();
```

Comment:
So before the loop, you do both PopActiveSnapshot() and CommitTransactionCommand(). Here you just Start a new transaction. Are you sure you don't need to PushActiveSnapshot(GetTransactionSnapshot()) as well?

Reply:
In my understanding you won't need a new snapshot since this is the last thing happening for this command. Since we block user transaction blocks, there cannot be any more operations going on after this that would require a new snapshot. At least that is my understanding. Also, this code exactly mirrors the code in the `cluster()` function in `src/backend/commands/cluster.c`, so AFAIK it is OK.